server/aesdsocket.c

```diff
@@ -0,0 +1,332 @@
+    if(fd[FD_DATA] == -1)
+	{
+		printf("ERROR: failed to open file, %s \n", FILE_OUT_PATH);
+		exit (EXIT_FAILURE);
+	}
+
+	sigemptyset(&sigSetOne);
+	sigaddset(&sigSetOne, SIGINT);
+	sigaddset(&sigSetOne, SIGTERM);
+
+    buffer_rx = (char *)malloc(BUFFER_SIZE*sizeof(char));
+    buffer_tx = (char *)malloc(BUFFER_SIZE*sizeof(char));
+	if(buffer_rx == NULL || buffer_tx == NULL)
+	{
+		printf("ERROR: failed to malloc buffers, exiting...\n");
+		exit (EXIT_FAILURE);
+	}	
+
+    if(daemonFlag)
+	{
+		pid = fork();
+
+		if(pid == -1)
+			exit (EXIT_FAILURE);
+
+		else if(pid != 0)
+			exit(0);
+
+		if(setsid() == -1)
+			exit (EXIT_FAILURE);
+
+		if(chdir("/") == -1)
+			exit (EXIT_FAILURE);
+
+		open("/dev/null", O_RDWR);
+
+		dup(0);
+		dup(0);
+	}
+	
+	while(1)
+	{
+		/* Get the accepted client fd */
+		fd[FD_CLIENT] = accept(fd[FD_SOCKET], (struct sockaddr *)&clientSockAddress, &addressSize); 
+
+		/* Error occurred in accept, exit (EXIT_FAILURE) on error */
+		if(fd[FD_CLIENT] == -1)
+		{
+			printf("ERROR: failed to accept client fd, exiting...\n");
+			free(buffer_rx);
+			exit (EXIT_FAILURE);
+		}
+
+		char *IP = inet_ntoa(clientSockAddress.sin_addr);
+		syslog(LOG_DEBUG, "Accepted connection from %s\n", IP);
+
+		/* Don't accept a signal while recieving data */
+		if((return_value = sigprocmask(SIG_BLOCK, &sigSetOne, &sigSetTwo)) == -1)
+			printf("ERROR: failed to sigprocmask, exiting...\n");
+		
+		/* Initially, write to starting of buffer */
+		bufferIndex = 0;	
+		
+		/* Receive data */
+		while((recievedByteCount = recv(fd[FD_CLIENT], buffer_rx + bufferIndex, BUFFER_SIZE, 0)) > 0)
+		{
+			/* Error occurred in recv, log error */
+			if(recievedByteCount == -1)
+			{
+				printf("ERROR: failed to received data successfully, exiting..., %s\n", strerror(errno));
+				exit (EXIT_FAILURE);
+			}
+
+			/* Detect newline character */
+			char* newlineloc = strchr(buffer_rx, '\n');
+
+			bufferIndex = bufferIndex + recievedByteCount;
+
+			/* Buffer size needs to be increased */
+			if((bufferIndex) >= (bufferCount * BUFFER_SIZE))
+			{
+				bufferCount++;
+				char* newptr = realloc(buffer_rx, bufferCount*BUFFER_SIZE*sizeof(char));
+
+				if(newptr == NULL)
+				{
+					free(buffer_rx);
+					printf("ERROR: failed to reallocate, exiting...\n");
+					exit (EXIT_FAILURE);
+				}
+
+				else buffer_rx = newptr;
+					
+			}
+
+			else if (newlineloc != NULL)
+			       	break;
+			
+		}
+		
+		// ********************* unmask signals that are pending...
+		if((return_value = sigprocmask(SIG_UNBLOCK, &sigSetTwo, NULL)) == -1)
+			printf("ERROR: failed sigprocmask, exiting...\n");
+		
+		// ********************* write to file
+		int wr_bytes = write(fd[FD_DATA], buffer_rx, bufferIndex);
+
+		if(wr_bytes != bufferIndex)
+			syslog(LOG_ERR, "Bytes written to file do not match bytes recieved from connection\n");
+	
+		filesize = lseek(fd[FD_DATA], 0, SEEK_CUR);
+		
+		char* newptr = realloc(buffer_tx, filesize*sizeof(char));
```

Comment:
@cchoi22915 what assumption does this make about the size of the RAM heap relative to the total size of all packets received?